### ISSUE TITLE
Weird and unhelpful assertion output for JSON

### ISSUE THREAD

**Issue by rcdailey** [fluentassertions/fluentassertions#1554]
### Description

When I am writing a unit test that involves using `BeEquivalentTo()` to compare two objects each with a member that holds a `JObject` from Newtonsoft.Json, the assertion output I get is not very helpful:

> Expected member Json[specifications][0][1][0][0] to be a collection with 3 item(s), but {{{empty}}, {{empty}}}"
"contains 1 item(s) less than"
"{{{empty}}, {{empty}}, {{empty}}}.

### Complete minimal example reproducing the issue

Create any NUnit test fixture class and insert the below method inside.

```cs
        [Test]
        public void MinimumRepro()
        {
            var thingToTest = new
            {
                Json = JObject.Parse(@"{
  'name': 'created',
  'specifications': [{
    'name': 'spec5',
    'fields': [{
      'name': 'value',
      'value': 'value2'
    }]
  }]
}")
            };

            var expectedThing = new
            {
                Json = JObject.Parse(@"{
  'name': 'created',
  'specifications': [{
    'name': 'spec5',
    'fields': [{
      'name': 'value',
      'value': 'value2',
      'newfield': 'newval'
    }]
  }]
}")
            };

            thingToTest.Should().BeEquivalentTo(expectedThing);
        }
    }
```

### Expected behavior:

I'd like the Json-specific information in the assertion to be clearly shown and formatted so that it can help me trace down the issue. As it stands right now, the `{{empty}}` values do not help :(

### Versions

```xml
    <PackageReference Update="FluentAssertions.Json" Version="5.*" />
    <PackageReference Update="FluentAssertions" Version="5.*" />
```

.NET 5

**Comment by rcdailey** [comment]
Please excuse my idiocy, I simply forgot this (I saw `Newtonsoft.Json` and misread it as `FluentAssertions.Json`):

```
using FluentAssertions.Json;
```

Output looks much better now. I apologize for the false alarm. Closing this one.

**Comment by rcdailey** [comment]
Sorry once more, I actually changed my test case and mislead myself. Let me clarify.

When I do the comparison like this (*and* with `FluentAssertions.Json` included!), the assertion looks fine:

```cs
thingToTest.Json.Should().BeEquivalentTo(expectedThing.Json);
```

But if I change it back to the original example, like below, I still get the bad output:

```cs
thingToTest.Should().BeEquivalentTo(expectedThing);
```

Obviously my brain is spinning here, sorry for the self-confusion.

**Comment by rcdailey** [comment]
Toying around with this some more, is this the way to get this behavior scoped to a single test fixture? So far this seems to work:

```cs

            AssertionOptions.AssertEquivalencyUsing(op => op
                .Using<JToken>(ctx => ctx.Subject.Should().BeEquivalentTo(ctx.Expectation))
                .WhenTypeIs<JToken>());

            thingToTest.Should().BeEquivalentTo(expectedThing);
```

**Comment by dennisdoomen** [comment]
When you include `FluentAssertion.Json`, which is in a [different repo](https://github.com/fluentassertions/fluentassertions.json), you're actually doing a proper JSON equivalency check. But if you use `FluentAssertion`, FA will do a recursive comparison on object properties and fields.

**Comment by rcdailey** [comment]
Not exactly. As I described in this thread, I expect the JSON equivalency check to automatically kick in when a property in an object graph is visited and it is a JSON type. I'm sure there's reasons why it already isn't doing this today, but that's definitely what I expect. Sometimes it's not feasible to break out an object graph to compare these items directly. I think `AssertionOptions` is a viable solution *except* that I can't limit options to a specific test fixture, as far as I know. And since I allow all of my test fixtures and test methods to run in parallel, this further complicates any alternative solution I could think of. It gets redundant adding the same 3 lines (`Using()` .. `WhenTypeIs()`) to every assertion.

**Comment by rcdailey** [comment]
There's no documentation for `IEquivalencyStep`. I think that's its own issue, but for now is this correct? If not, how would you suggest it be implemented?

```cs
    public class JsonEquivalencyStep : IEquivalencyStep
    {
        public bool CanHandle(IEquivalencyValidationContext context, IEquivalencyAssertionOptions config)
        {
            return context.Subject.GetType().IsAssignableTo(typeof(JToken));
        }

        public bool Handle(IEquivalencyValidationContext context, IEquivalencyValidator parent,
            IEquivalencyAssertionOptions config)
        {
            ((JToken) context.Subject).Should()
                .BeEquivalentTo((JToken) context.Expectation, context.Because, context.BecauseArgs);
            return true;
        }
    }
```

And using it I assume is like:

```cs
thingToTest.Should().BeEquivalentTo(expectedThing, op => op.Using(new JsonEquivalencyStep()));
```

**Comment by dennisdoomen** [comment]
> There's no documentation for `IEquivalencyStep`. I think that's its own issue, but for now is this correct? If not, how would you suggest it be implemented?

There is, but arguably not at a place where you expect it. Check it out the extensibility guide at https://fluentassertions.com/extensibility/#equivalency-assertion-step-by-step

But yes, exactly as in your example.

**Comment by rcdailey** [comment]
Ah, sorry I missed it. I really wish the search function in the documentation worked. I have never seen it return any results. Thank you for pointing in the right direction.

My biggest concern was the downcasting I had to do in the `Handle` method. I was worried I might be missing a form of this that used generics to avoid the casting. But seems not.

Thank you for your help. I'll close this issue.
